<launch>
    <!-- size of map, change the size inflate x, y, z according to your application -->
    <arg name="map_size_x"/>
    <arg name="map_size_y"/>
    <arg name="map_size_z"/>
    <arg name="init_x"/>
    <arg name="init_y"/>
    <arg name="init_z"/>
    <arg name="init_x1"/>
    <arg name="init_y1"/>
    <arg name="init_z1"/>    
    <arg name="init_x2"/>
    <arg name="init_y2"/>
    <arg name="init_z2"/>
    <arg name="init_x3"/>
    <arg name="init_y3"/>
    <arg name="init_z3"/>
    <arg name="init_x4"/>
    <arg name="init_y4"/>
    <arg name="init_z4"/>

    <arg name="point0_x"/>
    <arg name="point0_y"/>
    <arg name="point0_z"/>
    <arg name="point1_x"/>
    <arg name="point1_y"/>
    <arg name="point1_z"/>
    <arg name="point2_x"/>
    <arg name="point2_y"/>
    <arg name="point2_z"/>
    
    <arg name="target_x"/>
    <arg name="target_y"/>
    <arg name="target_z"/>
    <arg name="drone_id"/>
    <arg name="formation_type"/>
    <arg name="weight_obstacle"/>
    <arg name="weight_swarm"/>
    <arg name="weight_feasibility"/>
    <arg name="weight_sqrvariance"/>
    <arg name="weight_time"/>
    <arg name="weight_formation"/>
    <arg name="obstacle_clearance"/>
    <arg name="swarm_clearance"/>
    <arg name="replan_trajectory_time"/>
    
    <!-- topic of your odometry such as VIO or LIO -->
    <arg name="odom_topic"/>
    <!-- number of moving objects -->
    <!-- <arg name="obj_num" value="8" /> -->
    <!-- main algorithm params -->
    <include file="$(find ego_planner)/launch/advanced_param_lift.xml">
        <arg name="drone_id"                value="$(arg drone_id)"/>
        <!-- <arg name="formation_type"          value="$(arg formation_type)"/> -->
        <arg name="map_size_x_"             value="$(arg map_size_x)"/>
        <arg name="map_size_y_"             value="$(arg map_size_y)"/>
        <arg name="map_size_z_"             value="$(arg map_size_z)"/>
        <!-- <arg name="weight_obstacle"         value="$(arg weight_obstacle)"/>
        <arg name="weight_swarm"            value="$(arg weight_swarm)"/>
        <arg name="weight_feasibility"      value="$(arg weight_feasibility)"/>
        <arg name="weight_sqrvariance"      value="$(arg weight_sqrvariance)"/>
        <arg name="weight_time"             value="$(arg weight_time)"/>
        <arg name="weight_formation"        value="$(arg weight_formation)"/> -->
        <!-- <arg name="obstacle_clearance"      value="$(arg obstacle_clearance)"/> -->
        <!-- <arg name="swarm_clearance"         value="$(arg swarm_clearance)"/> -->
        <!-- <arg name="replan_trajectory_time"  value="$(arg replan_trajectory_time)"/> -->

        <!-- camera pose: transform of camera frame in the world frame -->
        <!-- depth topic: depth image, 640x480 by default -->
        <!-- don't set cloud_topic if you already set these ones! -->
        <arg name="odometry_topic" value="$(arg odom_topic)"/>
        <!-- <arg name="obj_num_set" value="$(arg obj_num)" /> -->
        <arg name="camera_pose_topic" value="pcl_render_node/camera_pose"/>
        <arg name="depth_topic" value="pcl_render_node/depth"/>
        <!-- topic of point cloud measurement, such as from LIDAR  -->
        <!-- don't set camera pose and depth, if you already set this one! -->
        <arg name="cloud_topic" value="pcl_render_node/cloud"/>
        <!-- intrinsic params of the depth camera -->
        <arg name="cx" value="321.04638671875"/>
        <arg name="cy" value="243.44969177246094"/>
        <arg name="fx" value="387.229248046875"/>
        <arg name="fy" value="387.229248046875"/>
        <!-- maximum velocity and acceleration the drone will reach -->
        <arg name="max_vel" value="1.5" />
        <arg name="max_acc" value="2.0" />
        <!--always set to 1.5 times grater than sensing horizen-->
        <arg name="planning_horizon" value="15" />
        <arg name="use_distinctive_trajs" value="false" />

    </include>
    <!-- trajectory server -->
    <node pkg="ego_planner" name="drone_$(arg drone_id)_traj_server" type="traj_server" output="screen">
        <remap from="position_cmd" to="drone_0_planning/pos_cmd"/>
        <remap from="~planning/trajectory" to="drone_0_planning/trajectory"/>

        <remap from="position_cmd1" to="drone_1_planning/pos_cmd"/>
        <remap from="~planning/trajectory1" to="drone_1_planning/trajectory"/>

        <remap from="position_cmd2" to="drone_2_planning/pos_cmd"/>
        <remap from="~planning/trajectory2" to="drone_2_planning/trajectory"/>

        <remap from="position_cmd3" to="drone_3_planning/pos_cmd"/>
        <remap from="~planning/trajectory3" to="drone_3_planning/trajectory"/>

        <remap from="position_cmd4" to="drone_4_planning/pos_cmd"/>
        <remap from="~planning/trajectory4" to="drone_4_planning/trajectory"/>

        <rosparam file="$(find ego_planner)/config/our_lift.yaml"/>

        <param name="traj_server/time_forward" value="1.0" type="double"/>
    </node>
    <!-- use simulator -->
    <include file="$(find ego_planner)/launch/simulator.xml">
        <arg name="drone_id" value="$(arg drone_id)"/>
        <arg name="map_size_x_" value="$(arg map_size_x)"/>
        <arg name="map_size_y_" value="$(arg map_size_y)"/>
        <arg name="map_size_z_" value="$(arg map_size_z)"/>
        <arg name="init_x_" value="$(arg init_x)"/>
        <arg name="init_y_" value="$(arg init_y)"/>
        <arg name="init_z_" value="$(arg init_z)"/>
        <!-- heli1 -->
        <arg name="init_x_1" value="$(arg init_x1)"/>
        <arg name="init_y_1" value="$(arg init_y1)"/>
        <arg name="init_z_1" value="$(arg init_z1)"/>
        <!-- heli2 -->
        <arg name="init_x_2" value="$(arg init_x2)"/>
        <arg name="init_y_2" value="$(arg init_y2)"/>
        <arg name="init_z_2" value="$(arg init_z2)"/>
        <!-- heli3 -->
        <arg name="init_x_3" value="$(arg init_x3)"/>
        <arg name="init_y_3" value="$(arg init_y3)"/>
        <arg name="init_z_3" value="$(arg init_z3)"/>
        <!-- heli4 -->
        <arg name="init_x_4" value="$(arg init_x4)"/>
        <arg name="init_y_4" value="$(arg init_y4)"/>
        <arg name="init_z_4" value="$(arg init_z4)"/>

        <arg name="odometry_topic" value="$(arg odom_topic)" />
    </include>
</launch>